File: IPWorks Encrypt Samples/EzRand/cpp/ezrand.cpp
/*
 * IPWorks Encrypt 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Encrypt in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksencrypt
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include "../../include/ipworksencrypt.h"
#define LINE_LEN 100
#define MESSAGE_LEN 1024

int main(int argc, char **argv)
{
	
  int ret_code = 0;
  char algorithm[LINE_LEN];
  char minValue[LINE_LEN];
  char maxValue[LINE_LEN];
  char seed[LINE_LEN];
  char bytesLen[LINE_LEN];
  char command[LINE_LEN];
  
  EzRand ezrand;

  printf("*****************************************************************************\n");
  printf("* This demo show how to get random numbers and bytes using algorithms like  *\n");
  printf("* ISAAC, MS Crypto API, etc.                                                *\n");
  printf("*****************************************************************************\n");

  while (true)
  {
    printf("\nSpecify an algorithm:\n");
    printf("1) ISAAC \n");
    printf("2) MS Crypto API \n");
    printf("3) Platform \n");
    printf("\nAlgorithm: ");
    fgets(algorithm, LINE_LEN, stdin);
    algorithm[strlen(algorithm)-1] = '\0';
    switch(atoi(algorithm)) 
    {
    case 1 : 
      ezrand.SetAlgorithm(RA_ISAAC);
      break;
    case 2 :
      ezrand.SetAlgorithm(RA_MSCRYPTO_API);
      break;
    case 3 :
      ezrand.SetAlgorithm(RA_PLATFORM);
      break;
    default :
      printf("\nInvalid algorithm.");
      continue;
    }

  ask:
    printf("\nGet random Integers(i) or Bytes(b)? ");
    fgets(command, LINE_LEN, stdin);

    if (tolower(command[0]) == 'i') //int
    {
      printf("\nMin Value [0]: ");
      fgets(minValue, LINE_LEN, stdin);
      minValue[strlen(minValue)-1] = '\0';
      if (strlen(minValue) == 0) strcat(minValue,"0");
      ret_code = ezrand.SetMin(atoi(minValue));
      if (ret_code) goto done;

      printf("\nMax Value [100]: ");
      fgets(maxValue, LINE_LEN, stdin);
      maxValue[strlen(maxValue)-1] = '\0';
      if (strlen(maxValue) == 0) strcat(maxValue,"100");
      ret_code = ezrand.SetMax(atoi(maxValue));
      if (ret_code) goto done;

      printf("Seed: ");
      fgets(seed, LINE_LEN, stdin);
      seed[strlen(seed)-1] = '\0';
      ret_code = ezrand.SetSeed(seed, strlen(seed));
      if (ret_code) goto done;

      printf("\nGetting random integers...\n");
      printf("Press <enter> to get next int or q to quit\n\n");

      do
      {
        ret_code = ezrand.GetNextInt();
        if (ret_code) goto done;
        
        printf("%d\n", ezrand.GetRandInt());
      } while (tolower(fgets(command, LINE_LEN, stdin)[0]) != 'q'); 
    }
    else if (tolower(command[0]) == 'b') //byte
    {
      printf("\nBytes Length [16]: ");
      fgets(bytesLen, LINE_LEN, stdin);
      bytesLen[strlen(bytesLen)-1] = '\0';
      if (strlen(bytesLen) == 0) strcat(bytesLen,"16");
      ret_code = ezrand.SetRandBytesLength(atoi(bytesLen));
      if (ret_code) goto done;

      printf("Seed: ");
      fgets(seed, LINE_LEN, stdin);
      seed[strlen(seed)-1] = '\0';
      ret_code = ezrand.SetSeed(seed, strlen(seed));
      if (ret_code) goto done;

      printf("\nGetting %d random bytes...\n", ezrand.GetRandBytesLength());
      printf("Press <enter> to get next bytes or q to quit\n\n");

      do 
      {
        ret_code = ezrand.GetNextBytes();
        if (ret_code) goto done;

        char *randBytes;
        int len;
        ret_code = ezrand.GetRandBytes(randBytes, len);
        if (ret_code) goto done;

        for (int i = 0; i < len; i++) 
        {
          printf("%02X ", (unsigned char)randBytes[i]);
        }
        printf("\n");
      } while (tolower(fgets(command, LINE_LEN, stdin)[0]) != 'q');
    }
    else {
      printf("Invalid Option");
      continue;
    }

    printf("\nDo you wish to generate another set of random integers or bytes? (y/n) ");
    if (tolower(fgets(command, LINE_LEN, stdin)[0] != 'y')) 
      break;
  }

done: 
  if (ret_code) // Got an error.  The user is done.
  {
    printf("\nError %d", ret_code);
    if (strlen(ezrand.GetLastError()) > 0) 
      printf(": %s\n", ezrand.GetLastError());
  }

	fprintf(stderr, "\n\npress <return> to continue...");
	getchar();
	exit(ret_code);
	return 0;
}

File: IPWorks Encrypt Samples/EzRand/php/ezrand.php
<?php
/*
 * IPWorks Encrypt 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Encrypt in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksencrypt
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworksencrypt_ezrand.php');
require_once('../include/ipworksencrypt_const.php');
?>
<?php

if ($argc < 3) {
  echo "usage: php ezrand.php [-i -min min -max max] [-c count] [-l length] [-s seed] -alg algorithm\n\n";
  echo "  -i           whether to generate random integers instead of bytes (optional)\n";
  echo "  -min         the lower bound of the random number to be generated (inclusive) (optional, default 0)\n";
  echo "  -max         the upper bound of the random number to be generated (exclusive) (optional, default 100)\n";
  echo "  -c           the number of random integers or byte arrays to generate\n";
  echo "  -l           the length of the byte array to be generated (optional, default 16)\n";
  echo "  -s           the seed to use (optional)\n";
  echo "  -alg         the random number algorithm to use, chosen from {ISAAC, CryptoAPI, Platform, RC4Random}\n";
  echo "\nExample: php ezrand.php -i -min 0 -max 1000 -c 5 -alg ISAAC\n";
  return;
}

try{
  $ezrand = new IPWorksEncrypt_Ezrand();
  $count = 1;
  $ints = false;

  for ($i = 1; $i < $argc; $i++) {
    if (str_starts_with($argv[$i],"-")) {
      if ($argv[$i] == "-alg") {
        switch (strtolower($argv[$i + 1])) {
          case "isaac":
            $ezrand->setAlgorithm(0);
            break;
          case "cryptoapi":
            $ezrand->setAlgorithm(1);
            break;
          case "platform":
            $ezrand->setAlgorithm(2);
            break;
          case "rc4random":
            $ezrand->setAlgorithm(4);
            break;
          default:
            echo "Invalid algorithm selection.\n";
            return;
        }
      }
      if ($argv[$i] == "-c") {
        $count = $argv[$i + 1];
      }
      if ($argv[$i] == "-min") {
        $ezrand->setMin($argv[$i + 1]);
      }
      if ($argv[$i] == "-max") {
        $ezrand->setMax($argv[$i + 1]);
      }
      if ($argv[$i] == "-s") {
        $ezrand->setSeed($argv[$i + 1]);
      }
      if ($argv[$i] == "-l") {
        $ezrand->setRandBytesLength($argv[$i + 1]);
      }
      if ($argv[$i] == "-i") {
        $ints = true;
      }
    }
  }

  if ($ints) {
    for ($i = 0; $i < $count; $i++) {
      $ezrand->doGetNextInt();
      echo $ezrand->GetRandInt() . "\n";
    }
  } else {
    $ezrand->doConfig("OutputEncoding=2"); // hex
    for ($i = 0; $i < $count; $i++) {
      $ezrand->doGetNextBytes();
      echo $ezrand->GetRandBytes() . "\n";
    }
  }
}  catch (Exception $e) {
  echo "Error: " . $e->getMessage() . "\n";
}
?>